src/adapter/src/coord/sequencer/inner/peek.rs

```diff
@@ -825,31 +834,80 @@ impl Coordinator {
     }
 
     #[tracing::instrument(level = "debug", skip_all)]
-    fn peek_stage_copy_to(
+    async fn peek_stage_copy_to_dataflow(
         &mut self,
-        ctx: &mut ExecuteContext,
+        ctx: ExecuteContext,
+        root_otel_ctx: OpenTelemetryContext,
         PeekStageCopyTo {
-            validity: _,
+            validity,
             optimizer,
             global_lir_plan,
         }: PeekStageCopyTo,
-    ) -> Result<ExecuteResponse, AdapterError> {
-        let session = ctx.session_mut();
+    ) {
+        let sink_id = global_lir_plan.sink_id();
+        let cluster_id = optimizer.cluster_id();
+
+        // Endpoints for sending and receiving responses.
+        let (tx, rx) = oneshot::channel::<CopyToResponse>();
+        let active_copy_to = ActiveCopyTo {
+            user: ctx.session().user().clone(),
+            conn_id: ctx.session().conn_id().clone(),
+            sender: Some(tx),
+            cluster_id,
+            depends_on: validity.dependency_ids.clone(),
+            dropping: false,
+        };
+        // Add metadata for the new COPY TO.
+        self.add_active_copy_to(sink_id, active_copy_to);
+
         let (df_desc, df_meta) = global_lir_plan.unapply();
 
-        self.emit_optimizer_notices(&*session, &df_meta.optimizer_notices);
+        self.emit_optimizer_notices(ctx.session(), &df_meta.optimizer_notices);
 
-        // Ship the dataflow
-        self.controller
-            .active_compute()
-            .create_dataflow(optimizer.cluster_id(), df_desc)
-            .unwrap_or_terminate("cannot fail to create dataflows");
-
-        // TODO(mouli): implement!
-        Err(AdapterError::Internal(format!(
-            "COPY TO '{}' is not yet implemented",
-            optimizer.copy_to_uri()
-        )))
+        // Ship dataflow.
+        self.ship_dataflow(df_desc, cluster_id).await;
+        let internal_cmd_tx = self.internal_cmd_tx.clone();
+        // Awaiting the response in a separate task to unblock
+        // further processing.
+        mz_ore::task::spawn(|| "copy_to await response", async move {
+            let response = match rx.await {
+                Ok(res) => Some(res),
+                Err(error) => {
+                    warn!("error receiving response back from sink: {:?}", error);
+                    None
+                }
+            };
+            let _ = internal_cmd_tx.send(Message::PeekStageReady {
+                ctx,
+                otel_ctx: root_otel_ctx,
+                stage: PeekStage::CopyToFinish(PeekStageCopyToFinish { validity, response }),
+            });
+        });
```

Comment:
I'm less familiar with the adapter codebase than I used to be, but this task seems potentially unnecessary to me.

Instead of having the `ActiveCopyTo` have a `sender` that sends back to this task, which then creates a new peek stage, which then finally retires the `ExecuteContext` ... can you instead just put the `ExecuteContext` inside of `ActiveCopyTo`, and have `ActiveCopyTo` call `ctx.retire` as appropriate?

Reply:
that works (and it simplified the code)!